src/amltk/optimization/optimizers/hebo.py

```diff
@@ -0,0 +1,272 @@
+
+    @classmethod
+    def create(
+        cls,
+        *,
+        space: Node | DesignSpace,
+        metrics: Metric | Sequence[Metric],
+        seed: Seed | None = None,
+        bucket: PathBucket | str | Path | None = None,
+        **optimizer_kwargs: Any,
+    ) -> HEBOOptimizer:
+        """Create an optimizer from HEBO.
+
+        Args:
+            space: The space to search over
+            metrics: The metrics to optimize.
+
+                * If `Metric`, then this is a single objective optimization with `HEBO`.
+                * If `Sequence[Metric]`, then this is a multi-objective optimization
+                    with `GeneralBO`.
+
+            seed: The seed to use for trials generated from this optimizer.
+            bucket: The bucket to store results of individual trials from this
+                optimizer.
+            **optimizer_kwargs: Keyword arguments to pass to the optimizer constructed.
+
+        Returns:
+            A HEBOOptimizer.
+        """
+        # TODO: Since hebo in it's observe will ignore anything that's inf, we can
+        # not have metrics that have an unbounded best, as these would get ignored...
+        # Probably need to raise an issue.
+        # Not really sure how to report or handle that case though as it's only a
+        # theoretical problem.
+        _check_metrics = [metrics] if isinstance(metrics, Metric) else metrics
+        if any(np.isinf(metric.optimal.value) for metric in _check_metrics):
+            raise ValueError(
+                "HEBO doesn't support metrics with an unbounded optimal value i.e. inf",
+            )
+
+        scramble_seed = amltk.randomness.as_int(seed)
+
+        if isinstance(bucket, str | Path):
+            bucket = PathBucket(bucket)
+
+        space = space if isinstance(space, DesignSpace) else parser(space)
+
+        match metrics:
+            case Metric() | [Metric()]:
+                optimizer = HEBO(
+                    space=space,
+                    scramble_seed=scramble_seed,
+                    **optimizer_kwargs,
+                )
+            case Sequence():
+                assert len(metrics) > 1
+                optimizer = GeneralBO(
+                    space=space,
+                    num_obj=len(metrics),
+                    # TODO: Not really sure if I should give a ref point or not,
+                    # especially if there are unbounded metrics.
+                    ref_point=np.array(cls.worst_possible_cost(metrics)),
+                    **optimizer_kwargs,
+                )
+
+        return cls(optimizer=optimizer, metrics=metrics, bucket=bucket, seed=seed)
+
+    @override
+    @classmethod
+    def preferred_parser(cls) -> HEBOParser:
+        return parser
+
+    @overload
+    @classmethod
+    def worst_possible_cost(cls, metric: Metric) -> float:
+        ...
+
+    @overload
+    @classmethod
+    def worst_possible_cost(cls, metric: Sequence[Metric]) -> list[float]:
+        ...
```

Comment:
May I ask if the "..." operator allows execution to simply fall through to the method defined below?

Reply:
The `...` is essentially just a value, just like writing the number `4`. For example,
These are both perfectly valid python code:

```python

def f(x: int):
    4
```

```python
def f(x: int):
    z = ...
```

What's making it fall through is the `@overload`. You can basically think of them as just doc strings that type checkers and language servers can use for better "intellisense". In the example you pointed out, if you pass in a `Sequence[Metric]`, you can expect `worst_possible_cost` to give you back a `list[float]`. If you pass in a `Metric`, you'll get back a `float`.

Without them, your editor/language server will say, `x = worst_possible_cost(single_metric)` is either a `float` or `list[float]`, even though you know it's a `float`.

Reply:
Thank you for the clarification.